src/libmongoc/src/mongoc/mongoc-stream-tls-openssl.c

```diff
@@ -564,31 +564,25 @@ _mongoc_stream_tls_openssl_handshake (mongoc_stream_t *stream,
    BIO_get_ssl (openssl->bio, &ssl);
 
    if (BIO_do_handshake (openssl->bio) == 1) {
-#if (OPENSSL_VERSION_NUMBER >= 0x10002000L)
-      X509 *peer = NULL;
-
-      if (tls->ssl_opts.allow_invalid_hostname) {
-         RETURN (true);
-      }
+      *events = 0;
 
-      peer = SSL_get_peer_certificate (ssl);
-      if (peer && (X509_check_host (peer, host, 0, 0, NULL) == 1 ||
-          X509_check_ip_asc (peer, host, 0) == 1)) {
-         X509_free (peer);
-         RETURN (true);
+#ifdef MONGOC_ENABLE_OCSP_OPENSSL
+      /* Validate OCSP */
+      if (openssl->ocsp_opts &&
+          1 != _mongoc_ocsp_tlsext_status (ssl, openssl->ocsp_opts)) {
```

Comment:
This is no longer done in the callback. I observed on OpenSSL 1.0.2/1.0.1 that against some servers, the callback would be unable to obtain the peer cert (i.e. `SSL_get_peer_certificate` returned NULL).

So instead, like the hostname check, this now does the OCSP check after the handshake.

FWIW one advantage of the callback is that when the callback returns a failure, that alerts the server the reason for the failure. Doing the OCSP check outside of the handshake means a failure would just close the connection client side. That does not seem like a big loss.